src/vm/threadsuspend.cpp

```diff
@@ -7411,8 +7411,417 @@ void STDCALL OnHijackStructInRegsWorker(HijackArgs * pArgs)
+    {
+#ifdef _DEBUG
+        BOOL GCOnTransition = FALSE;
+        if (g_pConfig->FastGCStressLevel()) {
+            GCOnTransition = GC_ON_TRANSITIONS(FALSE);
+        }
+#endif // _DEBUG
+
+#ifdef TIME_SUSPEND
+        g_SuspendStatistics.cntHijackTrap++;
+#endif // TIME_SUSPEND
+
+        CommonTripThread();
+#ifdef _DEBUG
+        if (g_pConfig->FastGCStressLevel()) {
+            GC_ON_TRANSITIONS(GCOnTransition);
+        }
+#endif // _DEBUG
+
+        // Update the references in the returned struct
+        *(size_t*)&pArgs->ReturnValue[0] = (size_t)ptr[0];
+        *(size_t*)&pArgs->ReturnValue[1] = (size_t)ptr[1];
+    }
+    GCPROTECT_END(); // trashes ptrs here!
+
+    frame.Pop();
+#else
+    PORTABILITY_ASSERT("OnHijackInteriorPointer2Worker not implemented on this platform.");
+#endif  // HIJACK_NONINTERRUPTIBLE_THREADS
+}
+
+void OnHijackObjectInteriorPointerWorker(HijackArgs * pArgs)
+{
+    CONTRACTL{
+        THROWS;
+    GC_TRIGGERS;
+    SO_TOLERANT;
+    } CONTRACTL_END;
+
+#ifdef HIJACK_NONINTERRUPTIBLE_THREADS
+    Thread         *thread = GetThread();
+
+    OBJECTREF       oref(ObjectToOBJECTREF(*(Object **)&pArgs->ReturnValue[0]));
+    void* ptr = (void*)(pArgs->ReturnValue[1]);
+
+#ifdef FEATURE_STACK_PROBE
+    if (GetEEPolicy()->GetActionOnFailure(FAIL_StackOverflow) == eRudeUnloadAppDomain)
+    {
+        RetailStackProbe(ADJUST_PROBE(DEFAULT_ENTRY_PROBE_AMOUNT), thread);
+    }
+#endif // FEATURE_STACK_PROBE
+
+    CONTRACT_VIOLATION(SOToleranceViolation);
+
+    thread->ResetThreadState(Thread::TS_Hijacked);
+
+    // Fix up our caller's stack, so it can resume from the hijack correctly
+    pArgs->ReturnAddress = (size_t)thread->m_pvHJRetAddr;
+
+    // Build a frame so that stack crawling can proceed from here back to where
+    // we will resume execution.
+    FrameWithCookie<HijackFrame> frame((void *)pArgs->ReturnAddress, thread, pArgs);
+
+    GCPROTECT_BEGIN(oref)
+    GCPROTECT_BEGININTERIOR(ptr)
+    {
+#ifdef _DEBUG
+        BOOL GCOnTransition = FALSE;
+        if (g_pConfig->FastGCStressLevel()) {
+            GCOnTransition = GC_ON_TRANSITIONS(FALSE);
+        }
+#endif // _DEBUG
+
+#ifdef TIME_SUSPEND
+        g_SuspendStatistics.cntHijackTrap++;
+#endif // TIME_SUSPEND
+
+        CommonTripThread();
+#ifdef _DEBUG
+        if (g_pConfig->FastGCStressLevel()) {
+            GC_ON_TRANSITIONS(GCOnTransition);
+        }
+#endif // _DEBUG
+
+        // Update the references in the returned struct
+        *((OBJECTREF *)&pArgs->ReturnValue[0]) = oref;
+        *(size_t*)&pArgs->ReturnValue[1] = (size_t)ptr;
+    }
+    GCPROTECT_END();        // trashes ptr here!
+    GCPROTECT_END();        // trashes oref here!
+
+    frame.Pop();
+#else
+    PORTABILITY_ASSERT("OnHijackObjectInteriorPointerWorker not implemented on this platform.");
+#endif //  HIJACK_NONINTERRUPTIBLE_THREADS
+}
+
+inline void SwapReturnRegisters(HijackArgs * pArgs)
+{
+    size_t swap = pArgs->ReturnValue[0];
```

Comment:
This swapping of arguments and having number of flavors of `OnHijack*` looks pretty convoluted. 

Instead, it would be better to move the switch statement over `ReturnKind` into `HijackFrame::GcScanRoots` (new overrride of `Frame::GcScanRoots`), and have just one shared version of everything else. Something like:

```
switch (returnKind)
{
case RT_Scalar:
    // nothing to report
    break;
case RT_Object:
   (*fn)(dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, CHECK_APP_DOMAIN);
    break;
case RT_ByRef:
    PromoteCarefully(fn, dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, GC_CALL_INTERIOR | GC_CALL_CHECK_APP_DOMAIN);
    break;
...
case RT_Obj_ByRef:
   (*fn)(dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, CHECK_APP_DOMAIN);
   PromoteCarefully(fn, dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[1]), sc, GC_CALL_INTERIOR | GC_CALL_CHECK_APP_DOMAIN);
   break;
...
```